tripal/tripal
# Permanent Files

# New Feature

Issue #1122 

## Description


This PR adds the ability for a site admin to mark an uploaded file as permanent.  This way, files  that need to stay on the site without getting automatically removed will be persistent.   

This fix does the following:
1.  Allows for a value of -1 to be a set in the `tripal_expiration_files.expiration_date` table which indicates that the file should not ever expire.
2. Updates file page on the User's account page.
    - Puts the usage and limitations in a collapsible fieldset so the page isn't so overwhelming.
    - Adds a 'Make Permanent' link as an Action for a file.
    - Removes the height of the file browser and sets a max-height instead.
3. Adds additional arguments to the `tripal_reset_file_expiration` API call to allow for setting a file as permanent and for setting a hard number of days. These are optional for backwards compatibility.
4. Updates the CSS class names so they have a 'tripal-user' namespace to avoid future conflicts.
5. Adjusts the CSS to remove styling that isn't generic (i.e. red colored selected file, blue colored file names)

## Testing?





To test this.
1.  Upload several files. The easiest way to test this is to use the FASTA or GFF uploader just to get the file uploaded (but don't actually submit an importer job).  For testing, you can use the Citrus [demo files](https://tripal.readthedocs.io/en/latest/user_guide/example_genomics/genomes_genes.html) in the Tripal User's Guide.
2. After uploading several files, go to your user account page and click on the 'Files' tab.  You should see the files you just uploaded.  Click on each one, then mouse over of the action links to see what file ID they have.  You'll see the File ID in the URL of the link.
3. Click the 'Make Permanent' link on them all, they should be updated to never expire.
4. If you look at the `tripal_expiration_files` table then you'll see the files that should not be expired should have a `-1` as the expiration date.
5. Using Drush run this command:  ```drush php-eval 'tripal_expire_files()'```. The files should remain.
6. Now run this Drush command to set their expiration date to right now:  ```drush php-eval 'tripal_reset_file_expiration(X,FALSE,0)'```. Be sure to change the `X` to the file ID of the file you want to remove.  The zero in the 3rd argument will force the file to have an expiration date of today.  Now if you run the ```drush php-eval 'tripal_expire_files()'``` again the file that was marked for expiration should disappear.
7. Return to the File page on the User account.  Click a file that is marked as permanent and then click the link 'Renew Expiration'. This should return the file to having an expiration date.

## Screenshots (if appropriate):

Here's a view of the files page with one file marked as permanent and the other with an expiration date.

![image](https://user-images.githubusercontent.com/1719352/110404090-180b1900-8033-11eb-850e-ed9b9a76aa0a.png)



## Additional Notes (if any):
